src/runtime-rs/crates/kata/src/event_publisher.rs

```diff
@@ -0,0 +1,203 @@
+// Copyright (c) 2021 Alibaba Cloud
+// Copyright (c) 2021 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::io::Write;
+use std::process::{Command, Stdio};
+use std::sync::mpsc::{channel, Receiver, Sender};
+use std::sync::Arc;
+
+use containerd_shim_protos::events::task::TaskOOM;
+use protobuf::Message;
+use slog::{debug, error, info, warn};
+
+use agent_client::Agent;
+
+const TASK_OOMEVENT_TOPIC: &str = "/tasks/oom";
+
+macro_rules! sl {
+    () => {
+        slog_scope::logger().new(slog::o!("source" => "event_publisher"))
+    };
+}
+
+#[derive(Debug, thiserror::Error)]
+enum EventError {
+    #[error("errors from protobuf: {0}")]
+    ProtobufError(#[source] protobuf::ProtobufError),
+    #[error("failed to spawn event publisher command: {0}")]
+    SpawnChild(#[source] std::io::Error),
+    #[error("failed to open stdin")]
+    OpenStdin,
+    #[error("failed to write stdin: {0}")]
+    WriteStdin(#[source] std::io::Error),
+    #[error("failed to wait stdout: {0}")]
+    WaitStdout(#[source] std::io::Error),
+}
+
+struct EventData {
+    topic: String,
+    data: Vec<u8>,
+}
+
+impl EventData {
+    fn new(event: impl Message, topic: &str) -> std::result::Result<Self, EventError> {
+        let any_message = ::protobuf::well_known_types::Any {
+            type_url: event.descriptor().full_name().into(),
+            value: event.write_to_bytes().map_err(EventError::ProtobufError)?,
+            ..Default::default()
+        };
+        let data = any_message
+            .write_to_bytes()
+            .map_err(EventError::ProtobufError)?;
+
+        Ok(Self {
+            topic: topic.into(),
+            data,
+        })
+    }
+
+    fn new_oom(container_id: &str) -> std::result::Result<Self, EventError> {
+        let event = TaskOOM {
+            container_id: container_id.into(),
+            ..Default::default()
+        };
+        Self::new(event, TASK_OOMEVENT_TOPIC)
+    }
+}
+
+/// Event publisher to send events to the Containerd.
+///
+/// It will automatically forward all TaskOOM events from agent to Containerd.
+/// TODO: support grace shutdown.
```

Comment:
- s/grace/graceful/ ?
- The runtime will need the ability to shut down gracefully for tracing to work, so if this difference between the 2.x runtime and the 3.x runtime needs to be captured somewhere.